### DISCUSSION TITLE
dual pane

### DISCUSSION THREAD

**Discussion by ghost** [sayanarijit/xplr#15] [Ideas]
Hi, thanks for creating this. Looks great!

I would love dual pane functionality like in mc.
Would that even be possible in a project like this?

Maybe this can be done with multiple windows like in VIM?

The reason I like dual pane is because most of the time I work with 2 or more locations.
Copy, move, ... you have a from and a destination path. Switching between these 2 locations should be fast and easy.

When I have to use Windows, I use Total Commander. This also has tabs and bookmarks. 

<img width="627" alt="image" src="https://user-images.githubusercontent.com/10061147/113707932-e6907680-96e0-11eb-9ac6-23e3a477d073.png">

**Comment by sayanarijit** [comment]
Hey. Thanks for explaining your workflow. I never used multiple panes, so I'll need some more information before implementing the feature.

What would you prefer?
- Multiple panes
- Multiple tabs
- Markable location history
- xplr sessions with a common selection buffer

I would prefer the last one i.e. single xplr sessions but with a synced selection buffer because it's an ideal candidate for a plugin. It's probably supported right now, if configured correctly. However, your input will help me make a decision.

**Reply by sayanarijit** [reply]
Or another option is the ability to send/copy the selection list to different xplr sessions.

**Reply by ghost** [reply]
I would prefer multiple panes, but if that doesn't fits in your project I would completely understand. It's your project :-)
Maybe I'm the only user who likes this. I would wait and see if other users would find this useful.

I use the `tab` key to switch between panes. This is default in mc and others.
In Vim I also work like this. I use `tab` to switch between windows.
`nnoremap <tab>   <c-w>w`

Not sure wat you mean by selection buffer?

For example to copy a file in mc I can use F5.
When pressing F5 the selected files are the source and the other pane is the destination.

The steps in xplr (and most file eplorers) are different. First select some files, go to the destination, copy here

<img width="974" alt="image" src="https://user-images.githubusercontent.com/10061147/113722995-370fd000-96f1-11eb-9c8b-fb673871130d.png">

**Comment by sayanarijit** [comment]
How would you like to have `--clone` to clone an xplr session from the given `$XPLR_SESSION_PATH`?

This way, you can select the files -> spawn a new xplr (terminal) window cloning from the current session -> paste the files -> quit the new window -> back to the parent session. This way we get the benefits of multiple tabs/windows while maintaining immutability since each new spawned session will have a different state & history. We don't even need to spawn a new window, we can spawn sessions from the same window if we want to.

@maximbaz I'd like to get your input too.

Let me know if you need further clarification.

**Reply by max-baz** [reply]
This is an interesting idea, however knowing a couple of `mc` users I _believe_ (please correct me if I'm wrong) that their preferred workflow is not to (1) go to source, (2) select what to copy, (3) go to target and (4) paste, but it is (1) go to source in one pane, (2) go to target in another pane, (3) select what to copy and (4) paste.

What you are suggesting is interesting and perhaps will have its own uses, but for people who prefer dual pane modes I don't think this will fit their accustomed workflow.

I'm not sure either that tabs like in `nnn` will solve this.

Lets get the feedback from folks who do use *commander managers 🤔

**Reply by aflyingpumpkin** [reply]
A possible solution would be to give users the ability to copy and paste the current selection. From one session to another. I was able to get copying to work by putting:
```
          ctrl-v:
            help: copy selection
            messages:
              - BashExecSilently: |
                  xclip -sel clipboard < $XPLR_PIPE_SELECTION_OUT
```
inside of the default modes ``key_bindings`` but I could not find a way to allow the pasting of selection inside the config file

**Reply by sayanarijit** [reply]
Hey @maximbaz, @aflyingpumpkin  thanks for explaining the mc-like copy-paste behavior. I think we can use [this](https://github.com/sayanarijit/xplr/wiki/Hacks#spawn-multiple-sessions-in-different-windows) and [this hack](https://github.com/sayanarijit/xplr/wiki/Hacks#copy--paste-using-system-clipboard) to get something similar to mc.

Demo
[![xplr-sessions.gif](https://s3.gifyu.com/images/xplr-sessions.gif)](https://github.com/sayanarijit/xplr/wiki/Hacks#spawn-multiple-sessions-in-different-windows))
[![copy](https://user-images.githubusercontent.com/11632726/116497845-eb9eab00-a8c5-11eb-8a3a-eef41299a9a5.gif)](https://github.com/sayanarijit/xplr/wiki/Hacks#copy--paste-using-system-clipboard))

**Reply by max-baz** [reply]
Personally, I think it's good enough, but I'm biased as I'm not really used to `mc` and I very much value the hackability over being bound to specific workflows 🙂 I'll let the other guys speak their mind.

**Comment by sayanarijit** [comment]
And here's a plugin implementing support for multiple contexts within xplr: https://github.com/igorepst/context-switch.xplr

**Comment by sayanarijit** [comment]
And finally, we have [dual-pane.xplr](https://github.com/sayanarijit/dual-pane.xplr).
